File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/IntelliJProjectTreeViewFix.java
package org.jetbrains.bsp.bazel.server.sync;

import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
import io.vavr.collection.Array;
import io.vavr.collection.HashSet;
import io.vavr.collection.Iterator;
import io.vavr.collection.List;
import io.vavr.collection.Seq;
import io.vavr.collection.Set;
import io.vavr.control.Option;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import org.jetbrains.bsp.bazel.server.sync.model.Label;
import org.jetbrains.bsp.bazel.server.sync.model.Module;
import org.jetbrains.bsp.bazel.server.sync.model.SourceSet;
import org.jetbrains.bsp.bazel.server.sync.model.Tag;
import org.jetbrains.bsp.bazel.workspacecontext.TargetsSpec;
import org.jetbrains.bsp.bazel.workspacecontext.WorkspaceContext;

public class IntelliJProjectTreeViewFix {
  public Seq<Module> createModules(
      URI workspaceRoot, Seq<Module> modules, WorkspaceContext workspaceContext) {
    if (isFullWorkspaceImport(workspaceContext)) {
      return createWorkspaceRootModule(workspaceRoot, modules);
    } else {
      return createMultipleModules(workspaceContext, workspaceRoot, modules);
    }
  }

  private boolean isFullWorkspaceImport(WorkspaceContext workspaceContext) {
    return importTargetSpecs(workspaceContext).exists(s -> s.startsWith("//..."));
  }

  private Seq<Module> createWorkspaceRootModule(URI workspaceRoot, Seq<Module> modules) {
    var existingRootDirectories = resolveExistingRootDirectories(modules);
    if (existingRootDirectories.contains(workspaceRoot)) {
      return Array.empty();
    }

    var rootModule = syntheticModule("bsp-workspace-root", workspaceRoot);
    return Array.of(rootModule);
  }

  private Seq<Module> createMultipleModules(
      WorkspaceContext workspaceContext, URI workspaceRoot, Seq<Module> modules) {
    var existingRootDirectories = resolveExistingRootDirectories(modules);
    var expectedRootDirs = resolveExpectedRootDirs(workspaceContext, workspaceRoot);
    var workspaceRootPath = Paths.get(workspaceRoot);
    return expectedRootDirs.flatMap(
        root -> {
          if (existingRootDirectories.contains(root)) {
            return Option.none();
          }

          var relative = workspaceRootPath.relativize(Paths.get(root)).toString();
          var moduleName = relative + "-modules-root";
          return Option.some(syntheticModule(moduleName, root));
        });
  }

  private Set<URI> resolveExistingRootDirectories(Seq<Module> modules) {
    return modules.iterator().map(Module::sourceSet).flatMap(SourceSet::sourceRoots).toSet();
  }

  private Seq<URI> resolveExpectedRootDirs(WorkspaceContext workspaceContext, URI workspaceRoot) {
    var dirs = rootDirsFromTargetSpecs(workspaceContext, workspaceRoot).sorted();
    if (dirs.isEmpty()) return List.empty();
    return removeAlreadyIncludedSubdirs(dirs);
  }

  private Seq<URI> removeAlreadyIncludedSubdirs(Seq<String> dirs) {
    var buffer = new ArrayList<String>();
    var current = dirs.get(0);
    buffer.add(current);
    for (var path : dirs.iterator().drop(1)) {
      if (!path.startsWith(current)) {
        current = path;
        buffer.add(current);
      }
    }
    return Array.ofAll(buffer.stream().map(URI::create));
  }

  private Seq<String> rootDirsFromTargetSpecs(
      WorkspaceContext workspaceContext, URI workspaceRoot) {
    var root = Paths.get(workspaceRoot);
    return importTargetSpecs(workspaceContext)
        .map(s -> stripSuffixes(s, ":all", "...", "/"))
        .map(s -> stripPrefixes(s, "//"))
        .map(root::resolve)
        .filter(Files::exists)
        .map(Path::toUri)
        .map(URI::toString)
        .toArray();
  }

  private String stripSuffixes(String s, String... suffixes) {
    for (var suffix : suffixes) {
      if (s.endsWith(suffix)) {
        s = s.substring(0, s.length() - suffix.length());
      }
    }
    return s;
  }

  private String stripPrefixes(String s, String... prefixes) {
    for (var prefix : prefixes) {
      if (s.startsWith(prefix)) {
        s = s.substring(prefix.length());
      }
    }
    return s;
  }

  private Iterator<String> importTargetSpecs(WorkspaceContext workspaceContext) {
    return Option.of(workspaceContext.getTargets())
        .iterator()
        .flatMap(TargetsSpec::getValues)
        .map(BuildTargetIdentifier::getUri);
  }

  private Module syntheticModule(String moduleName, URI baseDirectory) {
    HashSet<URI> resources = HashSet.of(baseDirectory);
    return new Module(
        Label.from(moduleName),
        true,
        List.empty(),
        HashSet.empty(),
        HashSet.of(Tag.NO_BUILD),
        baseDirectory,
        new SourceSet(HashSet.empty(), HashSet.empty()),
        resources,
        HashSet.empty(),
        Option.none());
  }
}

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/IntelliJProjectTreeViewFix.kt
package org.jetbrains.bsp.bazel.server.sync

import ch.epfl.scala.bsp4j.BuildTargetIdentifier
import org.jetbrains.bsp.bazel.bazelrunner.BazelInfo
import org.jetbrains.bsp.bazel.server.sync.model.Label
import org.jetbrains.bsp.bazel.server.sync.model.Module
import org.jetbrains.bsp.bazel.server.sync.model.SourceSet
import org.jetbrains.bsp.bazel.server.sync.model.Tag
import org.jetbrains.bsp.bazel.workspacecontext.WorkspaceContext
import java.net.URI
import java.nio.file.Files
import java.nio.file.Paths
import kotlin.io.path.Path
import kotlin.io.path.name
import kotlin.io.path.toPath
import kotlin.io.path.listDirectoryEntries
import kotlin.io.path.isSymbolicLink
import kotlin.io.path.readSymbolicLink

class IntelliJProjectTreeViewFix(
    private val bazelPathsResolver: BazelPathsResolver,
    private val bazelInfo: BazelInfo
) {

    fun createModules(
        workspaceRoot: URI,
        modules: Sequence<Module>,
        workspaceContext: WorkspaceContext
    ): Sequence<Module> {
        return if (isFullWorkspaceImport(workspaceContext, workspaceRoot)) {
            createWorkspaceRootModule(workspaceRoot, modules)
        } else {
            createMultipleModules(workspaceContext, workspaceRoot, modules)
        }
    }

    private fun isFullWorkspaceImport(workspaceContext: WorkspaceContext, workspaceRoot: URI): Boolean {
        return importTargetSpecs(workspaceContext).any { it.startsWith("//...") || it.startsWith("@//...") } ||
                workspaceContext.dotBazelBspDirPath.value.parent.toUri() == workspaceRoot
    }

    private fun createWorkspaceRootModule(
        workspaceRoot: URI, modules: Sequence<Module>
    ): Sequence<Module> {
        val excludeDirs = computeSymlinksToExclude(workspaceRoot)

        val existingRootDirectories = resolveExistingRootDirectories(modules)
        if (existingRootDirectories.contains(workspaceRoot)) {
            return modules.map {
                if (it.sourceSet.sourceRoots.contains(workspaceRoot)) {
                    it.copy(outputs = it.outputs + excludeDirs)
                } else it
            }
        }
        val rootModule = syntheticModule("bsp-workspace-root", workspaceRoot, excludeDirs)
        return modules + sequenceOf(rootModule)
    }

    private fun computeSymlinksToExclude(workspaceRoot: URI): Set<URI> {
        val stableSymlinkNames = setOf("bazel-out", "bazel-testlogs", "bazel-bin")
        val execRoot = Path(bazelInfo.execRoot)
        val execRootSymlinkNames = execRoot.name.let { name ->
            // newer bazel versions put workspace name as last component of exec root
            // still, the symlink to this directory is in form of bazel-<sanitized-workspace-name>
            // it seems to not accept '_' as a character, and it is replaced with '-'
            //
            // In older bazel versions, the exec root directory is named __main__ and I am using
            // less efficient heuristic: finding all bazel-* files that are symlinks that point to
            // exec root
            if (name != "__main__") {
                setOf("bazel-" + name.replace("[^A-Za-z0-9]".toRegex(), "-"))
            } else {
                workspaceRoot.toPath()
                    .listDirectoryEntries("bazel-*")
                    .filterNot { it.name in stableSymlinkNames }
                    .filter { it.isSymbolicLink() && it.readSymbolicLink() == execRoot }
                    .map { it.name }
            }
        }

        return (stableSymlinkNames + execRootSymlinkNames).map {
            workspaceRoot.toPath().resolve(it).toUri()
        }.toSet()
    }

    private fun createMultipleModules(
        workspaceContext: WorkspaceContext, workspaceRoot: URI, modules: Sequence<Module>
    ): Sequence<Module> {
        val existingRootDirectories = resolveExistingRootDirectories(modules)
        val expectedRootDirs = resolveExpectedRootDirs(workspaceContext, workspaceRoot)
        val workspaceRootPath = Paths.get(workspaceRoot)
        return modules + expectedRootDirs.mapNotNull { root: URI ->
            if (existingRootDirectories.contains(root)) {
                return@mapNotNull null
            }
            val relative = workspaceRootPath.relativize(Paths.get(root)).toString()
            val moduleName = "$relative-modules-root"
            syntheticModule(moduleName, root)
        }
    }

    private fun resolveExistingRootDirectories(modules: Sequence<Module>): Sequence<URI> =
        modules.map(Module::sourceSet).flatMap { it.sourceRoots }.distinct()

    private fun resolveExpectedRootDirs(
        workspaceContext: WorkspaceContext, workspaceRoot: URI
    ): Sequence<URI> {
        val dirs = rootDirsFromTargetSpecs(workspaceContext, workspaceRoot).sorted()
        return if (dirs.none()) emptySequence() else removeAlreadyIncludedSubdirs(dirs)
    }

    private fun removeAlreadyIncludedSubdirs(dirs: Sequence<String>): Sequence<URI> {
        var current = dirs.first() + "\n"
        return dirs.filter { path ->
            if (!path.startsWith(current)) {
                current = path
                true
            } else false
        }.map(URI::create)
    }

    private fun rootDirsFromTargetSpecs(
        workspaceContext: WorkspaceContext, workspaceRoot: URI
    ): Sequence<String> {
        val root = Paths.get(workspaceRoot)
        return importTargetSpecs(workspaceContext).map { stripSuffixes(it, ":all", "...", "/") }
            .map { stripPrefixes(it, "//") }.map(root::resolve).filter(Files::exists)
            .map(bazelPathsResolver::resolveUri).map(URI::toString)
    }

    private fun stripSuffixes(s: String, vararg suffixes: String): String =
        suffixes.fold(s) { acc, suffix ->
            if (acc.endsWith(suffix)) acc.substring(0, acc.length - suffix.length)
            else acc
        }

    private fun stripPrefixes(s: String, vararg prefixes: String): String =
        prefixes.fold(s) { acc, prefix ->
            if (acc.startsWith(prefix)) acc.substring(prefix.length)
            else acc
        }

    private fun importTargetSpecs(workspaceContext: WorkspaceContext): Sequence<String> =
        workspaceContext.targets.values.map(BuildTargetIdentifier::getUri).asSequence()

    private fun syntheticModule(moduleName: String, baseDirectory: URI, outputs: Set<URI> = emptySet()): Module {
        val resources = hashSetOf(baseDirectory)
        return Module(
            label = Label(moduleName),
            isSynthetic = true,
            directDependencies = emptyList(),
            languages = emptySet(),
            tags = hashSetOf(Tag.NO_BUILD),
            baseDirectory = baseDirectory,
            sourceSet = SourceSet(emptySet(), emptySet()),
            resources = resources,
            outputs = outputs,
            sourceDependencies = emptySet(),
            languageData = null,
            environmentVariables = hashMapOf()
        )
    }
}
